alpha/veneer/semver/semver.go

```diff
@@ -0,0 +1,428 @@
+package semver
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"sort"
+
+	"github.com/blang/semver/v4"
+	"github.com/operator-framework/operator-registry/alpha/action"
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+	"github.com/operator-framework/operator-registry/alpha/property"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// read in all the semver I/O custom elements
+// for each of the (currently predefined) channels in there:
+//    for each of the bundle images listed:
+//       render the bundle image via action.Render.Run()                     map[SemverVeneerChannelNames][]declcfg.Bundle
+//       grab channel kind, bundle name, version properties from the bundle  map[SemverVeneerChannelNames]map[bundleName]semver.Version
+//       walk the
+//
+// for each of the SemverVeneerChannelNames
+
+// data passed into this module externally
+type Veneer struct {
+	Ref string
+}
+
+// IO structs -- BEGIN
+type semverVeneerBundleEntry struct {
+	Image string `json:"image,omitempty"`
+}
+
+type candidateBundles struct {
+	Bundles []semverVeneerBundleEntry `json:"bundles,omitempty"`
+}
+type fastBundles struct {
+	Bundles []semverVeneerBundleEntry `json:"bundles,omitempty"`
+}
+type stableBundles struct {
+	Bundles []semverVeneerBundleEntry `json:"bundles,omitempty"`
+}
+
+type semverVeneer struct {
+	Schema                string           `json:"schema"`
+	GenerateMajorChannels bool             `json:"generateMajorChannels,omitempty"`
+	GenerateMinorChannels bool             `json:"generateMinorChannels,omitempty"`
+	AvoidSkipPatch        bool             `json:"avoidSkipPatch,omitempty"`
+	Candidate             candidateBundles `json:"candidate,omitempty"`
+	Fast                  fastBundles      `json:"fast,omitempty"`
+	Stable                stableBundles    `json:"stable"`
```

Comment:
remove this req.  Instead, enforce that we function as long as at least one channel has a single bundle in it.